bindings/sftp/sftp.go

```diff
@@ -0,0 +1,314 @@
+package sftp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+
+	sftpClient "github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/knownhosts"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/components-contrib/metadata"
+	"github.com/dapr/kit/logger"
+	kitmd "github.com/dapr/kit/metadata"
+)
+
+const (
+	metadataRootPath = "rootPath"
+	metadataFileName = "fileName"
+)
+
+// Sftp is a binding for file operations on sftp server.
+type Sftp struct {
+	metadata   *sftpMetadata
+	logger     logger.Logger
+	sftpClient *sftpClient.Client
+}
+
+// sftpMetadata defines the sftp metadata.
+type sftpMetadata struct {
+	RootPath       string `json:"rootPath"`
+	FileName       string `json:"fileName"`
+	Address        string `json:"address"`
+	Username       string `json:"username"`
+	Password       string `json:"password"`
+	PrivateKey     []byte `json:"privateKey"`
+	HostPublicKey  []byte `json:"hostPublicKey"`
+	KnownHostsFile string `json:"knownHostsFile"`
+	InsecureSSL    bool   `json:"isInsecureSSL"`
+}
+
+type createResponse struct {
+	FileName string `json:"fileName"`
+}
+
+type listResponse struct {
+	FileName    string `json:"fileName"`
+	IsDirectory bool   `json:"isDirectory"`
+}
+
+func NewSftp(logger logger.Logger) bindings.OutputBinding {
+	return &Sftp{logger: logger}
+}
+
+func (sftp *Sftp) Init(_ context.Context, metadata bindings.Metadata) error {
+	m, err := sftp.parseMetadata(metadata)
+	if err != nil {
+		return fmt.Errorf("failed to parse metadata: %w", err)
+	}
+
+	var auth []ssh.AuthMethod
+	var hostKeyCallback ssh.HostKeyCallback
+
+	if m.InsecureSSL {
+		hostKeyCallback = ssh.InsecureIgnoreHostKey()
+	} else if len(m.KnownHostsFile) > 0 {
+		hostKeyCallback, err = knownhosts.New(m.KnownHostsFile)
+		if err != nil {
+			return fmt.Errorf("sftp binding error: read known host file error: %w", err)
+		}
+	} else if len(m.HostPublicKey) > 0 {
+		hostPublicKey, _, _, _, err := ssh.ParseAuthorizedKey(m.HostPublicKey)
+		if err != nil {
+			return fmt.Errorf("sftp binding error: parse host public key error: %w", err)
+		}
+
+		hostKeyCallback = ssh.FixedHostKey(hostPublicKey)
+	}
+
+	if hostKeyCallback == nil {
+		return fmt.Errorf("sftp binding error: no host validation method provided")
+	}
+
+	if len(m.PrivateKey) > 0 {
+		signer, err := ssh.ParsePrivateKey(m.PrivateKey)
```

Comment:
Does this support encrypted private keys? They are fairly common, and they have a password too.

Reply:
`golang.org/x/crypto/ssh` package supports it. I will implement.